File: client/lib/models/availability.dart
class Availability {
  final String id;
  final String gymId;
  final String date;
  final String startTime;
  final String endTime;
  final String userId;

  Availability({
    required this.id,
    required this.gymId,
    required this.date,
    required this.startTime,
    required this.endTime,
    required this.userId,
  });

  factory Availability.fromJson(Map<String, dynamic> json) {
    return Availability(
      id: json['id'],
      gymId: json['gym_id'],
      date: json['date'],
      startTime: json['start_time'],
      endTime: json['end_time'],
      userId: json['user_id'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'gym_id': gymId,
      'date': date,
      'start_time': startTime,
      'end_time': endTime,
      'user_id': userId,
    };
  }
}

class NewAvailability {
  final int gymId;
  final String startTime;
  final String endTime;
  final String date;

  NewAvailability({
    required this.gymId,
    required this.startTime,
    required this.endTime,
    required this.date,
  });

  Map<String, dynamic> toJson() {
    return {
      'gymId': gymId,
      'startTime': startTime,
      'endTime': endTime,
      'date': date,
    };
  }
}

File: backend/schedule_service/src/models/availability.rs
use async_graphql::SimpleObject;
use chrono::{NaiveDateTime, Utc};
use diesel::RunQueryDsl;
use diesel::{deserialize::Queryable, prelude::Insertable, PgConnection, Selectable};

use serde::{Deserialize, Serialize};

use crate::schema::availabilities;

#[derive(Queryable, Serialize, Deserialize, SimpleObject, Selectable)]
#[diesel(table_name = crate::schema::availabilities)]
pub struct Availability {
    pub id: i32,
    pub user_id: i32,
    pub gym_id: i32,
    pub start_time: NaiveDateTime,
    pub end_time: NaiveDateTime,
    pub created_at: NaiveDateTime,
    pub updated_at: NaiveDateTime,
}

#[derive(Deserialize, Insertable)]
#[diesel(table_name = availabilities)]
pub struct NewAvailability {
    pub user_id: i32,
    pub gym_id: i32,
    pub start_time: NaiveDateTime,
    pub end_time: NaiveDateTime,
    pub created_at: NaiveDateTime,
    pub updated_at: NaiveDateTime,
}

impl NewAvailability {
    pub fn new(
        user_id: i32,
        gym_id: i32,
        start_time: NaiveDateTime,
        end_time: NaiveDateTime,
    ) -> Self {
        let now = Utc::now().naive_utc();
        NewAvailability {
            user_id,
            gym_id,
            start_time,
            end_time,
            created_at: now,
            updated_at: now,
        }
    }
}

impl Availability {
    pub fn create(new_availability: NewAvailability, conn: &mut PgConnection) -> Self {
        diesel::insert_into(availabilities::table)
            .values(&new_availability)
            .get_result(conn)
            .expect("Error creating availability")
    }
}
